src/roswire/proxy/roslaunch/ros2launch.py

```diff
@@ -0,0 +1,189 @@
+# -*- coding: utf-8 -*-
+__all__ = ('ROS2LaunchManager',)
+
+from typing import Collection, List, Mapping, Optional, Sequence, Tuple, Union
+import os
+import shlex
+
+from loguru import logger
+import attr
+import dockerblade
+
+from .config import LaunchConfig
+from .controller import ROSLaunchController
+from ... import exceptions as exc
+from .app.app import App
+
+
+@attr.s(eq=False)
+class ROS2LaunchManager:
+    """Provides access to `roslaunch <wiki.ros.org/roslaunch/>`_ for an
+    associated ROS system. This interface is used to locate, read, and write
+    `launch XML files <http://wiki.ros.org/roslaunch/XML>`_,
+    and to launch ROS nodes using those files.
+    """
+    _shell: dockerblade.shell.Shell = attr.ib(repr=False)
+    _files: dockerblade.files.FileSystem = attr.ib(repr=False)
+
+    def read(self,
+             filename: str,
+             *,
+             package: Optional[str] = None,
+             argv: Optional[Sequence[str]] = None
+             ) -> LaunchConfig:
+        """Produces a summary of the effects of a launch file.
+
+        Parameters
+        ----------
+        filename: str
+            The name of the launch file, or an absolute path to the launch
+            file inside the container.
+        package: str, optional
+            The name of the package to which the launch file belongs.
+        argv: Sequence[str], optional
+            An optional sequence of command-line arguments that should be
+            supplied to :code:`roslaunch`.
+
+        Raises
+        ------
+        PackageNotFound
+            If the given package could not be found.
+        LaunchFileNotFound
+            If the given launch file could not be found in the package.
+        """
+    raise NotImplementedError("ROS2 might not be able to read")
+
+    def write(self,
+              config: LaunchConfig,
+              *,
+              filename: Optional[str] = None
+              ) -> str:
+        """Writes a given launch configuration to disk as an XML launch file.
+
+        Parameters
+        ----------
+        config: LaunchConfig
+            A launch configuration.
+        filename: str, optional
+            The name of the file to which the configuration should be written.
+            If no filename is given, a temporary file will be created. It is
+            the responsibility of the caller to ensure that the temporary file
+            is appropriately destroyed.
+
+        Returns
+        -------
+        str
+            The absolute path to the generated XML launch file.
+        """
+    raise NotImplementedError("ROS2 might not be able to write")
+
+    def locate(self,
+               filename: str,
+               *,
+               app: App,
```

Comment:
Ah, now I see what you mean :slightly_smiling_face: So, the fix here is for there to be an `_app_instance: AppInstance = attr.ib()` attribute at the top. We can then remove `_shell` and `_files` and replace references to those with `_app_instance.shell` and `_app_instance.files`.